### PR TITLE
eve: adds dns fields for eve alert's json output

### PR DESCRIPTION
**Pull request by fooinha**
eve: adds dns fields for eve alert's json output

dns: uniformize json output following the same format as in alert event type

Check #2213 for more details.

### CODE REVIEW

**Comment by fooinha** [comment]
Hi, I may have another PR that code will conflict with this one. It's about payload printable size option for EVE alert json.

Advice please.

Thanx

**Comment by jasonish** [comment]
Proper review might take a little longer than usual.  I believe there is some work-in-progress going on with the DNS logging that may involve a more compressed output format.  And there will need to be some review and consensus on what that format should look like, which should be nailed down before adding DNS info to alerts.

Some of the development team is busy with training this week, then there will be some catch up to do.

File: src/output-json-alert.c [hunk 1]
```diff
@@ -281,6 +299,19 @@ static int AlertJson(ThreadVars *tv, JsonAlertLogThread *aft, const Packet *p)
             }
         }
 
+        if (json_output_ctx->flags & LOG_JSON_DNS) {
+            if (p->flow != NULL) {
+                FLOWLOCK_RDLOCK(p->flow);
```
Comment (inliniac) [inline src/output-json-alert.c]: This lock is no longer necessary now that https://github.com/inliniac/suricata/pull/2235 is merged.

**Comment by inliniac** [comment]
Could you have  a look at @decanio 's work in #2267 ?

File: src/output-json-alert.c (hunk 1 above)
Comment (fooinha) [inline src/output-json-alert.c]: I'll remove locks and try to rebase to master HEAD. Thanx.

**Comment by fooinha** [comment]
Applied feature and commit message from @decanio's work in #2267 .
Alert event from a rule with DNS protocol is not filtered and style of output  remains with unified output only ( so we don't generate multiple alert events for the same DNS request/response ) .

**Comment by fooinha** [comment]
Added some unit tests.

Test OutputJsonDnsQueryDiscreteTest01 -- Tests discrete query     : pass
Test OutputJsonDnsResponseUnifiedTest02 -- Tests unified response : pass

**Comment by fooinha** [comment]
Output:


- https://gist.github.com/fooinha/657b4ab5f6e5cffb4bacb2a45250f346#file-slit-json

- https://gist.github.com/fooinha/657b4ab5f6e5cffb4bacb2a45250f346#file-unified-json

**Comment by inliniac** [comment]
I like these outputs. @jasonish @decanio thoughts?

**Comment by decanio** [comment]
I haven't been following this thread closely.  Is the objective of this PR to add DNS metadata to alerts that trigger on DNS flows, and do it in a format that is compatible with my compact DNS PR?

If so the examples look ok to me.

**Comment by jasonish** [comment]
I like them to.  But I think when logging DNS details into an event it should only log the "unified" format, as we have to fit it in a single record anyways. It also means I think this PR has a dependence on #2275.

**Comment by fooinha** [comment]
Hi, the goal of this PR was only to add dns info into the alert event. 
The alert event contained the query/response in an unified style.

@inliniac , pointed to @decanio work in #2267 and #2275, and I included those ideas implementation in this PR.
 In the process I changed a bit the internal functions for dns ouput, and added 3 unit tests.

Yes, when alerting we force the "unified" format, so that we don't issue multiple events per rule.
Yes, the pasted gist locations, were files created by a running suricata with PR branch code.

Thanx.

**Comment by fooinha** [comment]
The closed PR #2213 has the initial part of the story.

**Comment by fooinha** [comment]
@inliniac , I saw your last comments at #2275.
This PR also implements the same, but with a significant rewrite for internal functions for dns json output.

Should we also tag this one with needsdocupdate ?

**Comment by inliniac** [comment]
I'm likely merging an updated version of #2275 for 3.2. This branch can then be rebased for the next major version. See https://lists.openinfosecfoundation.org/pipermail/oisf-devel/2016-October/003803.html

Btw, please have look at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Github_work_flow
Our workflow of github pull requests is to not modify them (and thus the branch they point to) after doing the PR. This PR is a nice example of why that is. Many unrelated commits make it messy. So closing.

**Comment by decanio** [comment]
I should have an update to my PR before the day is out.

**Comment by fooinha** [comment]
@decanio , do you need set of extra eyes to test it?

**Pull request closed without merging**